core/src/main/java/org/apache/iceberg/avro/BuildAvroProjection.java

```diff
@@ -154,16 +155,48 @@ public Schema.Field field(Schema.Field field, Supplier<Schema> fieldResult) {
 
   @Override
   public Schema union(Schema union, Iterable<Schema> options) {
-    Preconditions.checkState(AvroSchemaUtil.isOptionSchema(union),
-        "Invalid schema: non-option unions are not supported: %s", union);
-    Schema nonNullOriginal = AvroSchemaUtil.fromOption(union);
-    Schema nonNullResult = AvroSchemaUtil.fromOptions(Lists.newArrayList(options));
+    if (AvroSchemaUtil.isOptionSchema(union)) {
```

Comment:
Should this be using the `AvroSchemaWithType` visitor? I don't think that was written when this was added, but I like the idea of aligning the types with the right visitor, rather than having so much logic here.

Reply:
Are you suggesting refactoring `BuildAvroProjection` to use `AvroSchemaWithTypeVisitor` instead of `AvroCustomOrderSchemaVisitor`? If so, I agree. It will make projection implementation simpler. Do you think this can be done as a separate PR as it involves re-implementing logic out of scope of union support?